[Tan Xuan Yun] ip

tanxuanyun opened the pull request:
## Turtle
>“Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes)) 

Turtle frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,
1. download it from [here](https://github.com/tanxuanyun/ip/releases).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😆😆

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

tanxuanyun committed:
Level-1: Greet, Echo, Exit

tanxuanyun committed:
Level-2: Add, List

tanxuanyun committed:
Level-3: Mark as Done

tanxuanyun committed:
Level-4: ToDo, Event, Deadline

tanxuanyun committed:
A-TextUiTesting: Automated Text UI Testing

tanxuanyun committed:
Level-5: Handle Errors

tanxuanyun committed:
Level-6: Delete

tanxuanyun committed:
Save and load task to storage

tanxuanyun committed:
Add dates and times

tanxuanyun committed:
Merge branch 'branch-Level-8'

tanxuanyun committed:
Update date and time

tanxuanyun committed:
Update code with more OOP

tanxuanyun committed:
Add Gradle support

tanxuanyun committed:
Add JUnit tests to test the behavior of the Task.java and Todo.java

tanxuanyun committed:
Add MANIFEST.MF

tanxuanyun committed:
Add JavaDoc

tanxuanyun committed:
Add JavaDoc

tanxuanyun committed:
Update coding standard

tanxuanyun committed:
Add find task

tanxuanyun committed:
Merge branch 'branch-A-CodingStandard'

tanxuanyun committed:
Merge branch 'branch-Level-9'

tanxuanyun committed:
Edited JavaDoc

albertarielw commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,128 @@
+package duke;
```
First of all, I like how you extracted out the classes and packages, it makes the overall structure of your code really neat!

albertarielw commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,128 @@
+                            parts[1],
+                            LocalDateTime.parse(parts[2], dateTimeFormatter)
+                    );
+                    tasks.addTask(deadline);
+                    this.ui.printAddTask(deadline, this.tasks);
+                    break;
+                case "event":
+                    Task event = new Event(
+                            parts[1],
+                            LocalDateTime.parse(parts[2], dateTimeFormatter),
+                            LocalDateTime.parse(parts[3], dateTimeFormatter)
+                    );
+                    tasks.addTask(event);
+                    this.ui.printAddTask(event, this.tasks);
+                    break;
+                case "find":
+                    ArrayList<Task> targetTasks = this.tasks.findMatchTasks(parts[1]);
+                    this.ui.printMatchTask(targetTasks);
+                    break;
+                default:
+                    throw new DukeException("I'm sorry but I don't know what that means!");
+                }
+            } catch (DukeException e) {
+                this.ui.printDukeException(e);
+            } catch (DateTimeParseException e) {
+                this.ui.printDateTimeParseException();
+            }
+
+            current = this.ui.nextLine();
+        }
+
+        try {
+            this.storage.writeTasksToStorage(this.tasks);
+        } catch (IOException e) {
+            this.ui.printIoException(e);
+        }
+
+        this.ui.printDukeClosing();
+        this.ui.closeScanner();
+        System.exit(0);
```
I do not really prefer the switch statements approach as it makes the code really long. One suggestion is to make a class named command class to abstract out the cases perhaps?

albertarielw commented on src/main/java/duke/command/Parser.java:
```diff
@@ -0,0 +1,91 @@
+
+        String[] segments = input.split(" ");
+
+        if (input.equals("list")) {
+            return "list|";
+        } else if (segments[0].equals("find")) {
+            input = input.replace("find", "");
+            if (input.equals("")) {
+                throw new DukeException("The description of a todo cannot be empty.");
+            }
+            return "find|" + input.trim();
+        } else if (segments[0].equals("mark")) {
+            int taskId;
+            if (segments[1].equals("")) {
+                taskId = -1;
+            } else {
+                taskId = Integer.parseInt(segments[1]);
+            }
+            return "mark|" + taskId;
+        } else if (segments[0].equals("unmark")) {
+            int taskId;
+            if (segments[1].equals("")) {
+                taskId = -1;
+            } else {
+                taskId = Integer.parseInt(segments[1]);
+            }
+            return "unmark|" + taskId;
+        } else if (segments[0].equals("todo")) {
```
Perhaps it may be better to abstract out some of the logic so that the function is not too long? One rule of thumb i read is to keep function length to be a screen size at max

albertarielw commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,132 @@
+public class Storage {
+    private final File file;
+
+    /**
+     * Construct a new Storage object with a given file path.
+     * Ensures the file is accessible.
+     *
+     * @param filePath The path to the file to store data in.
+     * @throws IOException Handles case of File not being found or location not existing.
+     */
+    public Storage(String filePath) throws IOException {
+        String[] parts = filePath.split("/");
+
+        File folder = new File(parts[0]);
+        if (!folder.exists()) {
+            folder.mkdir();
+        }
+
+        File file = new File(filePath);
```
I like how your JavaDoc is very detailed and clear, especially in this class

albertarielw commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,92 @@
+package duke.task;
+import java.util.ArrayList;
+
+/**
+ * Encapsulates a list of tasks.
+ */
+public class TaskList {
+
+    private ArrayList<Task> tasks;
+
+    /**
+     * Constructs a new {@code TaskList} with no tasks.
```
I really like how this class is very succinct and has very detailed JavaDoc. Your code looks really neat here

albertarielw commented on src/main/java/duke/utilities/Ui.java:
```diff
@@ -0,0 +1,145 @@
+ */
+public class Ui {
+
+    private static String line = "--------------------";
```
One suggestion for your UI class is to abstract out the string values (perhaps enum class can even be used) and make the variables final as well as follow the naming convention for constants

for instance you can change

> private static String line = "--------------------";

to

> private static final String LINE = "--------------------";

albertarielw commented on text-ui-test/input.txt:
```diff
@@ -0,0 +1,17 @@
+todo read book
+deadline return book /by 2019-10-15 06:00
+event project meeting /from 2019-10-15 06:00 /to 2019-10-15 07:00
+todo join sports club
+todo borrow book
+mark 1
+mark 2
+mark 4
+unmark 4
+list
+todo
```
I like how your testcases are very detailed!

cliftonfelix commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,132 @@
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.Scanner;
```
It's nice that the ordering of import statements is consistent

cliftonfelix commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,132 @@
+     * @throws IOException Handles File not being found when writing to hard disk.
+     */
+    public void writeTasksToStorage(TaskList taskList) throws IOException {
+        ArrayList<Task> tasks = taskList.getTasks();
```
Nice plural variable

cliftonfelix commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,132 @@
+                String output = "T|" + task.getStatusIcon() + "|" + task.getDescription();
+                fileWriter.write(output);
+                break;
+            }
```
I like that there is no indentation for case clauses, but I don't think that the curly brackets are necessary after case clauses

cliftonfelix reviewed (commented):
Overall, in terms of coding standards, all is good. Just a few nits to fix.

tanxuanyun committed:
Added GUI

tanxuanyun committed:
Revert "Added GUI"

This reverts commit 15e510c7366d7097dc8acbefea4d1023cf900d32.

tanxuanyun committed:
Revert "Revert "Added GUI""

This reverts commit a011383d66b7fe27698d8092c31a849e055511d6.

tanxuanyun committed:
Add level 10 branch

tanxuanyun committed:
Add jar file

tanxuanyun committed:
Add assertions to check important assumptions

tanxuanyun committed:
Parses class: extract each command into functions

parsesCommand method is long.

parsesCommand does not follow SLAP.

Extracting the each command into to different function.

Help to increase readability in parsesCommand method.

tanxuanyun committed:
Add assertion to check important assumption

Tasklist class does not check if taskId and index are valid.

The invalid taskId and index will cause error.

Assertion is added to check if taskId and index are valid.

Ensure that error will not be cause by invalid taskId and index

tanxuanyun committed:
Merge pull request #1 from tanxuanyun/branch-A-Assertions

Branch A-Assertions

tanxuanyun committed:
Merge pull request #2 from tanxuanyun/branch-A-CodeQuality

Branch A-CodeQuality

tanxuanyun committed:
Set up Continuous Integration using GitHub Actions

tanxuanyun committed:
Add reminder for unmark deadline tasks

tanxuanyun committed:
Add representative screenshot and user guide

tanxuanyun committed:
Edit header comments

tanxuanyun committed:
Touch up code quality issues by following feedback

tanxuanyun committed:
Update Readme
